upstream: New original_dst cluster type.

jrajahalme opened the pull request:
Envoy already supports SO_ORIGINAL_DST to retrieve the original
destination IP address and L4 port number that were overwritten by
netfilter REDIRECT action.  So far the original destination address
has been only used to switch from the initial listener to another
listener on the original address.

This patch adds a new cluster type that uses the original
destination address as the upstream destination for the requests
coming from a connection where the SO_ORIGINAL_DST returns an address
that is different from the initial destination address the connection
was received on.  The request must be routed to this new cluster type
for this to happen, and other cluster types can still be used as
before, ignoring the original destination address on the upstream
side.

The "original_dst" cluster type makes transparent proxying with envoy
easier in scenarios where it is impractical to configure a dedicated
cluster for each possible upstream host, e.g., when acting as an
egress proxy facing a large external network.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme commented:
Please ignore the first patch, it's format fixes that I need locally to pass check format fix, but probably need to be removed before merging.

jrajahalme commented:
For some reason GitHub shows the last two patches reversed above. f251b30 is the last patch of the series.

jrajahalme commented:
Rebased to fix issues with the first iteration of the PR. Sorry for the churn.

mattklein123 commented on include/envoy/http/filter.h:
```diff
@@ -71,9 +71,9 @@ class StreamFilterCallbacks {
-   * @return uint64_t the ID of the originating connection for logging purposes.
+   * @return const Network::Connection* the originating connection.
    */
-  virtual uint64_t connectionId() PURE;
```
Please leave around the connectionId() function and mark it for deprecation. We will remove it in 1.5.0 release cycle.

mattklein123 commented on include/envoy/network/connection.h:
```diff
@@ -162,6 +162,16 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
+  /**
+   * Set boolean telling if the connection is using an original destination address.
+   */
+  virtual void setUsingOriginalDst(bool using_original_dst) PURE;
```
I would remove this from the interface. It should be passed as part of the Impl constructor.

mattklein123 commented on source/common/upstream/cluster_manager_impl.h:
```diff
@@ -172,7 +172,15 @@ class ClusterManagerImpl : public ClusterManager {
+        auto host = lb_->chooseHost(context);
+        if (host) {
+          auto real_host = host->getRealHost(context);
+          if (real_host)
```
braces around all if statements

mattklein123 commented on source/common/upstream/original_dst_cluster.cc:
```diff
@@ -0,0 +1,83 @@
+    std::vector<Json::ObjectSharedPtr> hosts_json = config.getObjectArray("hosts");
+    n_hosts = hosts_json.size();
+  } catch (...) {
+  } // Catch missing hosts config.
```
? just use `hasObject` and fail if it has it.

mattklein123 replied:
Please don't catch (...) (and definitely don't catch and do nothing). Only specific exceptions. In this case per previous comment I would just use hasObject()

mattklein123 commented on include/envoy/upstream/thread_local_cluster.h:
```diff
@@ -3,6 +3,8 @@
 namespace Envoy {
 namespace Upstream {
 
+class LoadBalancerContext;
```
In general only forward declare if it causes a circular include. That might be the case here and elsewhere but just checking.

mattklein123 commented on source/common/upstream/original_dst_cluster.h:
```diff
@@ -0,0 +1,95 @@
+ * A new 'real host' is created for each destination, but currently
+ * stats and other state refers to the logical host.
+ */
+class OriginalDstCluster : public ClusterImplBase {
```
The major problem I see with this approach (which is not quite the same as LogicalDnsCluster) is that you have the potential to add an unlimited # of hosts (and by association connection pools) that are never going to get cleaned up once they are created. At minimum I think we would need stats specific to this cluster type telling us how many real hosts we have. We don't specifically track the # of connection pools in stats for HTTP stuff, so it's going to be hard to tell exactly how many hosts/conn_pools we have created.

With that said, you can of course run out of memory in any number of other ways, and presumably you are using this cluster type in a specific deployment where you probably have control over what is happening, but I still feel that there should be some type of output telling us what is going on, even if the hosts aren't specifically in the primary cluster. 

Note also that the way this is implemented your "real hosts" can/will be duplicated across all worker threads. I'm not sure if this matters or not.

Stepping back a bit, there is another approach that could be taken (just sketching it out a bit):

- Write a filter that would sit inline to check to see if the original_dst is already in your cluster
- If not, pause iteration, and do worker <-> main thread sync to add it to cluster, then continue iteration.
- If it is, just continue, and do normal LB with some type of hint that picks a specific host in the load balancer by searching.

I haven't thought through the previous very fully, but if that interests you, we can discuss further (and maybe should do a quick video conference).

I guess given how self contained this change is (mostly) I don't have a huge objection to the approach as long as the downsides are understood and well documented. LMK what you think.

jrajahalme replied:
Thanks, exactly the kind of feedback I was hoping for :-)

> The major problem I see with this approach (which is not quite the same as LogicalDnsCluster) is that you have the potential to add an unlimited # of hosts (and by association connection pools) that are never going to get cleaned up once they are created. At minimum I think we would need stats specific to this cluster type telling us how many real hosts we have. We don't specifically track the # of connection pools in stats for HTTP stuff, so it's going to be hard to tell exactly how many hosts/conn_pools we have created.

I had assumed that dynamically created hosts would get cleared when last reference to them dies out, and as they are not currently held in any containers that should happen soon after the connection is terminated? In testing I noticed that connection get's terminated right after the response is received.

> With that said, you can of course run out of memory in any number of other ways, and presumably you are using this cluster type in a specific deployment where you probably have control over what is happening, but I still feel that there should be some type of output telling us what is going on, even if the hosts aren't specifically in the primary cluster.

Assuming that we keep the dynamically created Hosts out of the cluster HostSets, do you have any suggestions on how to add stats like this?

> Note also that the way this is implemented your "real hosts" can/will be duplicated across all worker threads. I'm not sure if this matters or not.

Maybe I should make the logical host hold references to the created real hosts and reuse them both in the same thread and across the threads. This approach needs a timer to clean out unused hosts after a while, though.

> Stepping back a bit, there is another approach that could be taken (just sketching it out a bit):

> Write a filter that would sit inline to check to see if the original_dst is already in your cluster
If not, pause iteration, and do worker <-> main thread sync to add it to cluster, then continue iteration.

You mean inline with the http filters? And this is due to existing infra allowing filter iteration to be paused whereas no such mechanism exists at the cluster output stage?

How about making the router filter do the pausing if the cluster tells so, maybe via the new cluster->chooseHost() I added in this series?

> If it is, just continue, and do normal LB with some type of hint that picks a specific host in the load balancer by searching.

This might be best accomplished by a specialized load balancer that is designed to work like this. original_dst_cluster configuration could then be limited to that LB.

> I haven't thought through the previous very fully, but if that interests you, we can discuss further (and maybe should do a quick video conference).

At this stage I'd like learn more about the need to synchronize with the main thread when the host sets of the clusters are updated. I had a hunch it would be needed, and hence did not look deeper into that yet, hoping to come up with a solution avoiding such synchronization need, as it adds latency and complexity.

Right now I think maybe still hiding the real hosts behind one logical host, but making them (explicitly) stick around for some time to be reused and more explicitly managed with some kind of stats.

> I guess given how self contained this change is (mostly) I don't have a huge objection to the approach as long as the downsides are understood and well documented. LMK what you think.

My aim was to keep the changes to the overall architecture minimal. I'll look into adding more explicit management/stats of the real hosts and see if we can still avoid global synchronization for that. I'll wait for you responses to gauge which direction to go from here, though.

Thanks!

mattklein123 commented:
> I had assumed that dynamically created hosts would get cleared when last reference to them dies out, and as they are not currently held in any containers that should happen soon after the connection is terminated? In testing I noticed that connection get's terminated right after the response is received.

For HTTP, unfortunately not exactly. (This is why this is way easier for the tcp_proxy case). For HTTP we end up creating thread local connection pools that are persistent and keyed by hosts. They effectively live forever until the host is explicitly removed from the cluster: https://github.com/lyft/envoy/blob/master/source/common/upstream/cluster_manager_impl.cc#L518

The way you have the code written now, connection pools will be created (per worker) and will live forever.

> Assuming that we keep the dynamically created Hosts out of the cluster HostSets, do you have any suggestions on how to add stats like this?

In this most basic case, your HostImpl could just inc a stat when it is created. As previously discussed, this will be per thread, and will never get cleaned up in the HTTP case. The more I think about it the more I'm not super thrilled with this overall approach.

> At this stage I'd like learn more about the need to synchronize with the main thread when the host sets of the clusters are updated. I had a hunch it would be needed, and hence did not look deeper into that yet, hoping to come up with a solution avoiding such synchronization need, as it adds latency and complexity. Right now I think maybe still hiding the real hosts behind one logical host, but making them (explicitly) stick around for some time to be reused and more explicitly managed with some kind of stats.

I would take a look at addMemberUpdateCb() in the code and how it is used in both the main thread and the worker threads inside of the cluster manager. With that said, this is some of the most complicated code in the server so I admit it will take some time to understand.

Adding continuation support when the router calls chooseHost() is possible, but I would really prefer to not make that code more complicated than it already is. It's already very complicated. I think a network level filter could be installed that would do something like the following (network level is good because it could be used for other cases also):

- Check of original_dst host is already in thread local target cluster. If so, do nothing else.
- If not, post() to the main thread to add a new host, and pause filter iteration.
- On main thread add host, potentially set timer for stale host cleanup, post back to filter thread to continue filter. (This dance will not be simple but is possible)
- Have special load balancer (IP target LB) that just returns the host by target IP that you put into the hint.

I think this would work and would be pretty clean. You would also see all hosts in /clusters output, etc. I think this is probably the right approach. Anyhow, this is all pretty complicated. If you want to have a meeting about this please email me at mklein@lyft.com and we can discuss.

jrajahalme commented:
New approach with explicit management of the original destination hosts via a specialized load balancer.

jrajahalme commented:
Not merged with current master yet, seeking feedback on the overall design.

jrajahalme commented:
Rebased, fixed format and and cleaned up, should build now. Apart from rebasing, the last commit has these changes:
- do not use a static pointer to a singleton OriginalDstCluster. This makes testing easier and removes the limit that only one OriginalDstCluster may be instantiated. Checks that enforce OriginalDstCluster::LoadBalancer to be always used only with OriginalDstCluster make the dynamic_cast from Cluster& to OriginalDstCluster& safe.
- break a for loop after erase(), which invalidates iterators. Continuation was not needed anyway, and continuing resulted into a segmentation fault when multiple hosts expired at the same time.
- More rigorous testing, also catching the bug described above.
- Only update members in the cleanup timeout if at least one host is removed.

I hope that this forced update, actually makes this easier to review as well..

mattklein123 commented on configs/original_dst_proxy.json:
```diff
@@ -0,0 +1,44 @@
+{
+    "listeners": [{
+        "address": "tcp://0.0.0.0:10000",
+	"use_original_dst": true,
```
nit: indent

mattklein123 commented on include/envoy/http/filter.h:
```diff
@@ -76,6 +76,11 @@ class StreamFilterCallbacks {
   /**
+   * @return const Network::Connection* the originating connection.
+   */
+  virtual const Network::Connection* connection() PURE;
```
Can we please bring back all the logging, etc. changes you did in the other PR and mark connectionId() method deprecated and add to DEPRECATED.md? I would like to delete that method in 1.5.0 release cycle. Best if you could actually split this part of the change into a separate PR first so we can get that in and make this review less lines?

mattklein123 commented on include/envoy/network/connection.h:
```diff
@@ -179,6 +179,12 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
   virtual uint32_t bufferLimit() const PURE;
+
+  /**
+   * boolean telling if the connection's local address is an original destination address,
```
nit: @return

mattklein123 commented on include/envoy/upstream/load_balancer.h:
```diff
@@ -18,9 +18,14 @@ class LoadBalancerContext {
+   * @return Optional<uint64_t> the optional hash key to use during load balancing.
    */
-  virtual const Optional<uint64_t>& hashKey() const PURE;
+  virtual Optional<uint64_t> hashKey() const PURE;
```
If possible please split this change into separate PR also.

mattklein123 commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -536,6 +537,12 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::ClusterEntry(
   }
+  case LoadBalancerType::OriginalDst: {
+    lb_.reset(new OriginalDstCluster::LoadBalancer(
+        host_set_, dynamic_cast<OriginalDstCluster&>(
```
please add comment on why this dynamic_cast is guaranteed to succeed.

mattklein123 commented on source/common/upstream/original_dst_cluster.h:
```diff
@@ -0,0 +1,71 @@
+  OriginalDstCluster(const Json::Object& config, Runtime::Loader& runtime, Stats::Store& stats,
+                     Ssl::ContextManager& ssl_context_manager, Event::Dispatcher& dispatcher,
+                     bool added_via_api);
+  ~OriginalDstCluster() {}
```
nit: del line, not needed

mattklein123 commented on source/common/upstream/upstream_impl.cc:
```diff
@@ -84,6 +85,13 @@ ClusterInfoImpl::ClusterInfoImpl(const Json::Object& config, Runtime::Loader& ru
     lb_type_ = LoadBalancerType::RingHash;
+  } else if (string_lb_type == "original_dst_lb") {
+    if (config.getString("type") != "original_dst") {
+      throw EnvoyException(
```
I realize we have been bad about doing this in many places, but please test with EXPECT_THROW_WITH_MESSAGE for better tests.

mattklein123 commented on source/common/upstream/original_dst_cluster.cc:
```diff
@@ -0,0 +1,163 @@
+                                     const std::vector<HostSharedPtr>& hosts_removed) -> void {
+    // Update the hosts map
+    for (auto host : hosts_removed) {
+      ENVOY_LOG(debug, "REMOVING HOST {}", host->address()->asString());
```
nit: please follow other logging which generally uses lower case (applies to all)

mattklein123 commented on source/common/upstream/original_dst_cluster.cc:
```diff
@@ -0,0 +1,163 @@
+  host_set_.addMemberUpdateCb([this](const std::vector<HostSharedPtr>& hosts_added,
+                                     const std::vector<HostSharedPtr>& hosts_removed) -> void {
+    // Update the hosts map
+    for (auto host : hosts_removed) {
```
`const auto&` (same below)

mattklein123 commented on source/common/upstream/original_dst_cluster.cc:
```diff
@@ -0,0 +1,163 @@
+HostConstSharedPtr
+OriginalDstCluster::LoadBalancer::chooseHost(const LoadBalancerContext* context) {
+  if (context) {
+    auto connection = context->downstreamConnection();
```
`const auto` just for clarity. Same in other places if possible.

mattklein123 commented on source/common/upstream/original_dst_cluster.cc:
```diff
@@ -0,0 +1,163 @@
+    auto connection = context->downstreamConnection();
+    // Verify that we are not looping back to ourselves!
+    if (connection && connection->usingOriginalDst()) {
+      auto& localAddr = connection->localAddress();
```
nit: all vars snake_case

mattklein123 commented on source/common/upstream/original_dst_cluster.cc:
```diff
@@ -0,0 +1,163 @@
+          // Add the new host to the map.  We just failed to find it in
+          // our local map above, so just insert it.
+          host_map_.insert({localAddr.asString(), host});
+          host->weight(2); // Mark as used.
```
?

htuch replied:
Yeah, also wondering about the magic 2 here.

mattklein123 commented on source/common/upstream/original_dst_cluster.cc:
```diff
@@ -0,0 +1,163 @@
+HostSharedPtr OriginalDstCluster::createHost(Network::Address::InstanceConstSharedPtr address) {
+  HostSharedPtr host(
+      new HostImpl(info_, info_->name() + address->asString(), address, false, 1, ""));
+  dispatcher_.post([this, host]() -> void { addHost(host); });
```
I need to think about this more (well, you do, but I'm helping 😉 ), but I think there is a race condition here with CDS cluster removal. I think it's possible that cluster has been removed on main thread by the time the post gets there. We can chat here or offline about potential solutions (and if there really is a problem).

mattklein123 commented on source/common/upstream/original_dst_cluster.cc:
```diff
@@ -0,0 +1,163 @@
+  std::vector<HostSharedPtr> to_be_removed;
+  auto hostset = hosts();
+  for (auto host : hostset) {
+    if (host->weight() == 1) {
```
I see you are somehow using weight for cleanup marking. Can you comment/explain that somewhere?

jrajahalme commented:
> On Jul 20, 2017, at 11:42 AM, Matt Klein <notifications@github.com> wrote:
> 
> @mattklein123 commented on this pull request.
> 
> In general, I like the approach. This works for me. Moving forward, can we:
> 
> Split out the things I mentioned into independent PRs. Those will be non-controversial and easy to merge.

Will do.
> Then we can iterate on this and finish it up (note that final PR will need docs and good explanation about all of this).
Right, there is a lot of polish to do (remove unnecessary debug messages, docs, coverage, etc.)
> Please make sure that all error cases are covered (maybe you already did that).
I did most of the coverage work earlier today :-)

> @htuch <https://github.com/htuch> or @dnoe <https://github.com/dnoe> can you draw straws and can one of you take a quick skim through this to make sure we are all on the same page?
> 
> Thanks @jrajahalme <https://github.com/jrajahalme> !
> 
> In configs/original_dst_proxy.json <https://github.com/lyft/envoy/pull/1246#discussion_r128593156>:
> 
> > @@ -0,0 +1,44 @@
> +{
> +    "listeners": [{
> +        "address": "tcp://0.0.0.0:10000",
> +	"use_original_dst": true,
> nit: indent
> 
> In include/envoy/http/filter.h <https://github.com/lyft/envoy/pull/1246#discussion_r128593430>:
> 
> > @@ -76,6 +76,11 @@ class StreamFilterCallbacks {
>    virtual uint64_t connectionId() PURE;
>  
>    /**
> +   * @return const Network::Connection* the originating connection.
> +   */
> +  virtual const Network::Connection* connection() PURE;
> Can we please bring back all the logging, etc. changes you did in the other PR and mark connectionId() method deprecated and add to DEPRECATED.md? I would like to delete that method in 1.5.0 release cycle. Best if you could actually split this part of the change into a separate PR first so we can get that in and make this review less lines?
> 
> 

I missed the existence of DEPRECATED.md and the special signage, will add. So keep the APIs, but already remove all internal use so that nothing breaks when the deprecated APIs are removed? No problem :-)

> 
> In include/envoy/network/connection.h <https://github.com/lyft/envoy/pull/1246#discussion_r128593558>:
> 
> > @@ -179,6 +179,12 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
>     * Get the value set with setBufferLimits.
>     */
>    virtual uint32_t bufferLimit() const PURE;
> +
> +  /**
> +   * boolean telling if the connection's local address is an original destination address,
> nit: @return <https://github.com/return>
> In include/envoy/upstream/load_balancer.h <https://github.com/lyft/envoy/pull/1246#discussion_r128594250>:
> 
> >     */
> -  virtual const Optional<uint64_t>& hashKey() const PURE;
> +  virtual Optional<uint64_t> hashKey() const PURE;
> If possible please split this change into separate PR also.
> 
> 

OK.
> 
> In source/common/upstream/cluster_manager_impl.cc <https://github.com/lyft/envoy/pull/1246#discussion_r128594648>:
> 
> > @@ -536,6 +537,12 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::ClusterEntry(
>                                         parent.parent_.random_));
>      break;
>    }
> +  case LoadBalancerType::OriginalDst: {
> +    lb_.reset(new OriginalDstCluster::LoadBalancer(
> +        host_set_, dynamic_cast<OriginalDstCluster&>(
> please add comment on why this dynamic_cast is guaranteed to succeed.
> 
OK.
> 
> In source/common/upstream/original_dst_cluster.h <https://github.com/lyft/envoy/pull/1246#discussion_r128594811>:
> 
> > +
> +namespace Envoy {
> +namespace Upstream {
> +
> +/**
> + * The OriginalDstCluster is a dynamic cluster that automatically adds hosts as needed based on the
> + * original destination address of the downstream connection.  These hosts are also automatically
> + * cleaned up after they have not seen traffic for a configurable cleanup interval time
> + * ("cleanup_interval_ms").
> + */
> +class OriginalDstCluster : public ClusterImplBase {
> +public:
> +  OriginalDstCluster(const Json::Object& config, Runtime::Loader& runtime, Stats::Store& stats,
> +                     Ssl::ContextManager& ssl_context_manager, Event::Dispatcher& dispatcher,
> +                     bool added_via_api);
> +  ~OriginalDstCluster() {}
> nit: del line, not needed
> 
> In source/common/upstream/original_dst_cluster.cc <https://github.com/lyft/envoy/pull/1246#discussion_r128595404>:
> 
> > +#include "common/network/address_impl.h"
> +#include "common/network/utility.h"
> +
> +namespace Envoy {
> +namespace Upstream {
> +
> +OriginalDstCluster::OriginalDstCluster(const Json::Object& config, Runtime::Loader& runtime,
> +                                       Stats::Store& stats,
> +                                       Ssl::ContextManager& ssl_context_manager)
> +    : ClusterImplBase(config, runtime, stats, ssl_context_manager) {
> +  size_t n_hosts = 0;
> +  try {
> +    std::vector<Json::ObjectSharedPtr> hosts_json = config.getObjectArray("hosts");
> +    n_hosts = hosts_json.size();
> +  } catch (...) {
> +  } // Catch missing hosts config.
> Please don't catch (...) (and definitely don't catch and do nothing). Only specific exceptions. In this case per previous comment I would just use hasObject()
> 
> 

Forgot about that, sorry.

> 
> In source/common/upstream/upstream_impl.cc <https://github.com/lyft/envoy/pull/1246#discussion_r128595965>:
> 
> > @@ -84,6 +85,13 @@ ClusterInfoImpl::ClusterInfoImpl(const Json::Object& config, Runtime::Loader& ru
>      lb_type_ = LoadBalancerType::Random;
>    } else if (string_lb_type == "ring_hash") {
>      lb_type_ = LoadBalancerType::RingHash;
> +  } else if (string_lb_type == "original_dst_lb") {
> +    if (config.getString("type") != "original_dst") {
> +      throw EnvoyException(
> I realize we have been bad about doing this in many places, but please test with EXPECT_THROW_WITH_MESSAGE for better tests.
> 
> 

Will look into this, haven’t worked with cluster manager tests yet.

> 
> In source/common/upstream/original_dst_cluster.cc <https://github.com/lyft/envoy/pull/1246#discussion_r128596605>:
> 
> > +#include <vector>
> +
> +#include "common/network/address_impl.h"
> +#include "common/network/utility.h"
> +
> +namespace Envoy {
> +namespace Upstream {
> +
> +OriginalDstCluster::LoadBalancer::LoadBalancer(HostSet& host_set, OriginalDstCluster& parent)
> +    : host_set_(host_set), parent_(parent) {
> +  // host_set_ is initially empty.
> +  host_set_.addMemberUpdateCb([this](const std::vector<HostSharedPtr>& hosts_added,
> +                                     const std::vector<HostSharedPtr>& hosts_removed) -> void {
> +    // Update the hosts map
> +    for (auto host : hosts_removed) {
> +      ENVOY_LOG(debug, "REMOVING HOST {}", host->address()->asString());
> nit: please follow other logging which generally uses lower case (applies to all)
> 
> 

Most of this debugging will be gone for the “non-RFC” PR, as we are now verifying this behavior with the tests.

> 
> In source/common/upstream/original_dst_cluster.cc <https://github.com/lyft/envoy/pull/1246#discussion_r128596923>:
> 
> > +#include <string>
> +#include <vector>
> +
> +#include "common/network/address_impl.h"
> +#include "common/network/utility.h"
> +
> +namespace Envoy {
> +namespace Upstream {
> +
> +OriginalDstCluster::LoadBalancer::LoadBalancer(HostSet& host_set, OriginalDstCluster& parent)
> +    : host_set_(host_set), parent_(parent) {
> +  // host_set_ is initially empty.
> +  host_set_.addMemberUpdateCb([this](const std::vector<HostSharedPtr>& hosts_added,
> +                                     const std::vector<HostSharedPtr>& hosts_removed) -> void {
> +    // Update the hosts map
> +    for (auto host : hosts_removed) {
> const auto& (same below)
> 
Good catch, no point taking a new reference here.

> 
> In source/common/upstream/original_dst_cluster.cc <https://github.com/lyft/envoy/pull/1246#discussion_r128597057>:
> 
> > +          ENVOY_LOG(debug, "ADDING HOST {}: EXISTED ALREADY!", host->address()->asString());
> +          found = true;
> +          break;
> +        }
> +      }
> +      if (!found) {
> +        host_map_.insert({host->address()->asString(), host});
> +      }
> +    }
> +  });
> +}
> +
> +HostConstSharedPtr
> +OriginalDstCluster::LoadBalancer::chooseHost(const LoadBalancerContext* context) {
> +  if (context) {
> +    auto connection = context->downstreamConnection();
> const auto just for clarity. Same in other places if possible.
> 
OK.

> 
> In source/common/upstream/original_dst_cluster.cc <https://github.com/lyft/envoy/pull/1246#discussion_r128597157>:
> 
> > +        }
> +      }
> +      if (!found) {
> +        host_map_.insert({host->address()->asString(), host});
> +      }
> +    }
> +  });
> +}
> +
> +HostConstSharedPtr
> +OriginalDstCluster::LoadBalancer::chooseHost(const LoadBalancerContext* context) {
> +  if (context) {
> +    auto connection = context->downstreamConnection();
> +    // Verify that we are not looping back to ourselves!
> +    if (connection && connection->usingOriginalDst()) {
> +      auto& localAddr = connection->localAddress();
> nit: all vars snake_case
> 
> 

OK.

> 
> In source/common/upstream/original_dst_cluster.cc <https://github.com/lyft/envoy/pull/1246#discussion_r128597239>:
> 
> > +        if (localIP) {
> +          Network::Address::InstanceConstSharedPtr hostIpPort;
> +          if (localIP->version() == Network::Address::IpVersion::v4) {
> +            hostIpPort.reset(
> +                new Network::Address::Ipv4Instance(localIP->addressAsString(), localIP->port()));
> +          } else {
> +            hostIpPort.reset(
> +                new Network::Address::Ipv6Instance(localIP->addressAsString(), localIP->port()));
> +          }
> +          HostSharedPtr host = parent_.createHost(hostIpPort);
> +
> +          ENVOY_LOG(debug, "CREATED HOST {}", host->address()->asString());
> +          // Add the new host to the map.  We just failed to find it in
> +          // our local map above, so just insert it.
> +          host_map_.insert({localAddr.asString(), host});
> +          host->weight(2); // Mark as used.
> ?
> 
> In source/common/upstream/original_dst_cluster.cc <https://github.com/lyft/envoy/pull/1246#discussion_r128597937>:
> 
> > +    std::vector<Json::ObjectSharedPtr> hosts_json = config.getObjectArray("hosts");
> +    n_hosts = hosts_json.size();
> +  } catch (...) {
> +  } // Catch missing hosts config.
> +  if (n_hosts != 0) {
> +    throw EnvoyException("original_dst clusters must have no hosts configured");
> +  }
> +
> +  cleanup_timer_->enableTimer(cleanup_interval_ms_);
> +}
> +
> +// Called by a worker thread.
> +HostSharedPtr OriginalDstCluster::createHost(Network::Address::InstanceConstSharedPtr address) {
> +  HostSharedPtr host(
> +      new HostImpl(info_, info_->name() + address->asString(), address, false, 1, ""));
> +  dispatcher_.post([this, host]() -> void { addHost(host); });
> I need to think about this more (well, you do, but I'm helping 😉 ), but I think there is a race condition here with CDS cluster removal. I think it's possible that cluster has been removed on main thread by the time the post gets there. We can chat here or offline about potential solutions (and if there really is a problem).
> 
> 

Good point. I’m thinking that if the CDS cluster removal is posted to the main thread, then all the queued posts would still have a valid this pointer. If that can’t be done then we should pass a cluster name here instead and do a cluster lookup when executing the posted function. On the other hand, it may be that there is no need to dynamically add a original_dst_cluster, as one can serve all destinations?

> 
> In source/common/upstream/original_dst_cluster.cc <https://github.com/lyft/envoy/pull/1246#discussion_r128598096>:
> 
> > +
> +// Called in the main thread.
> +void OriginalDstCluster::addHost(HostSharedPtr host) {
> +  HostVectorSharedPtr new_hosts(new std::vector<HostSharedPtr>(hosts()));
> +  new_hosts->emplace_back(host);
> +  updateHosts(new_hosts, createHealthyHostList(*new_hosts), empty_host_lists_, empty_host_lists_,
> +              {host}, {});
> +}
> +
> +void OriginalDstCluster::cleanup() {
> +  ENVOY_LOG(debug, "CLEANUP.");
> +  HostVectorSharedPtr new_hosts(new std::vector<HostSharedPtr>);
> +  std::vector<HostSharedPtr> to_be_removed;
> +  auto hostset = hosts();
> +  for (auto host : hostset) {
> +    if (host->weight() == 1) {
> I see you are somehow using weight for cleanup marking. Can you comment/explain that somewhere?
> 
> 

Just wanted to have something that works for the RFC PR. If using the weight like this is acceptable I can document it, an alternative would be to add another field like it for mark/sweep cleanup. Basically whenever a chooseHost chooses a host it marks it as “used” (== 2), and whenever the timer callback sweeps through the hosts it marks them as “potentially unused” (== 1). If the sweep sees a host marked with 1, it is known that the host has not been used for at least cleanup_interval_ms and we can remove the host.

  Jarno

mattklein123 commented:
> Good point. I’m thinking that if the CDS cluster removal is posted to the main thread, then all the queued posts would still have a valid this pointer. If that can’t be done then we should pass a cluster name here instead and do a cluster lookup when executing the posted function. On the other hand, it may be that there is no need to dynamically add a original_dst_cluster, as one can serve all destinations?

From recent discussions with @htuch on this we should try if at all possible to make things work dynamically if possible. In this case, I think it's not too hard. You can post by name and just drop the post on the main thread if name lookup doesn't resolve because it was deleted. However, there is still a tiny edge case in which name gets removed and re-added as something else. I realize the chance of this happening is basically zero, however as a purist we should make it correct. The way to do that is probably to add an incrementing ID for each cluster much like we do for listener for basically the same purpose. Then you can post name/id and make sure it matches.

> If using the weight like this is acceptable I can document it, an alternative would be to add another field like it for mark/sweep cleanup. 

Personally I would prefer adding an explicit mark/sweep boolean and leaving weight unused. It's kind of confusing IMO.

htuch commented:
+1 to dynamic where possible, even the filesystem based configs will be dynamic in v2. I'll take a look at this PR from the Google side tomorrow.

htuch commented on source/common/upstream/original_dst_cluster.cc:
```diff
@@ -0,0 +1,163 @@
+      for (auto it = range.first; it != range.second; ++it) {
+        if (it->second == host) {
+          ENVOY_LOG(debug, "ADDING HOST {}: EXISTED ALREADY!", host->address()->asString());
+          found = true;
```
Can we use `std::find_if` here?

htuch commented on source/common/upstream/original_dst_cluster.cc:
```diff
@@ -0,0 +1,163 @@
+        // Add a new host
+        auto localIP = localAddr.ip();
+        if (localIP) {
+          Network::Address::InstanceConstSharedPtr hostIpPort;
```
How come we use the local address below as the host? I thought the idea was to use the real address from original_dst? I'm probably missing something major, so apologies if this is a silly question. A comment in the code could also be beneficial to the reader on the rationale here.

htuch commented on source/common/json/config_schemas.cc:
```diff
@@ -1305,7 +1305,7 @@ const std::string Json::Schema::CLUSTER_SCHEMA(R"EOF(
       "type" : {
         "type" : "string",
-        "enum" : ["static", "strict_dns", "logical_dns", "sds"]
+        "enum" : ["static", "strict_dns", "logical_dns", "sds", "original_dst"]
```
Can we add docs on these new options in https://github.com/lyft/envoy/blob/master/docs/configuration/cluster_manager/cluster.rst etc.

jrajahalme commented:
> On Jul 20, 2017, at 5:21 PM, Matt Klein <notifications@github.com> wrote:
> 
> Good point. I’m thinking that if the CDS cluster removal is posted to the main thread, then all the queued posts would still have a valid this pointer. If that can’t be done then we should pass a cluster name here instead and do a cluster lookup when executing the posted function. On the other hand, it may be that there is no need to dynamically add a original_dst_cluster, as one can serve all destinations?
> 
> From recent discussions with @htuch <https://github.com/htuch> on this we should try if at all possible to make things work dynamically if possible. In this case, I think it's not too hard. You can post by name and just drop the post on the main thread if name lookup doesn't resolve because it was deleted. However, there is still a tiny edge case in which name gets removed and re-added as something else. I realize the chance of this happening is basically zero, however as a purist we should make it correct. The way to do that is probably to add an incrementing ID for each cluster much like we do for listener for basically the same purpose. Then you can post name/id and make sure it matches.
> 
> 

A few thoughts:

- Would it be bad if I changed the ClusterPtr type (which is a std::unique_ptr) to a ClusterSharedPtr? Then posts could take these and essentially hold a reference so that destruction is postponed until last of these dies off. Overall might be safer, and likely faster than passing around strings and doing hash lookup.

- ClusterInfo is already reference counted (std::shared_ptr), but it does not help here.

- If we have unique IDs, then why not have a map just from the unique id to the object? No strings attached :-)

  Jarno

jrajahalme commented:
> On Jul 21, 2017, at 7:17 AM, htuch <notifications@github.com> wrote:
> 
> @htuch commented on this pull request.
> 
> Looks great to me design wise, I can review the PRs that get split out from the Google side.
> 
> In configs/setup_original_dst_netns.sh <https://github.com/lyft/envoy/pull/1246#discussion_r128766860>:
> 
> > @@ -0,0 +1,23 @@
> +#
> What references these shell scripts for setup/teardown? Or are they intended to be run manually (in which case I would suggest a different location, or explicit pairing with a config). Might also make sense to put these in a distinct PR.
> 
> 

Nothing references them now, but they are intended as guides what needs to be done to test the original dst cluster on a host. They work together with the config, and I would have put the commands as comments to the json, but json takes no comments. Will go for explicit pairing with the config.

> 
> In configs/cleanup_original_dst_netns.sh <https://github.com/lyft/envoy/pull/1246#discussion_r128767592>:
> 
> > @@ -0,0 +1,15 @@
> +#
> +# Cleanup network namespace after testing Envoy original_dst cluster
> +#
> +NETNS=$1
> +TARGET_IP=$2
> +ENVOY_PORT=10000
> +
> +# remove iptables rule
> +iptables -t nat -D PREROUTING --src 0/0 --dst $TARGET_IP -p tcp --dport 80 -j REDIRECT --to-ports $ENVOY_PORT
> Nit: Wrap all variables derefs in "" as per style guide https://google.github.io/styleguide/shell.xml#Quoting <https://google.github.io/styleguide/shell.xml#Quoting>
> 

Will do.. learning as I go :-)

> 
> In configs/cleanup_original_dst_netns.sh <https://github.com/lyft/envoy/pull/1246#discussion_r128767742>:
> 
> > @@ -0,0 +1,15 @@
> +#
> +# Cleanup network namespace after testing Envoy original_dst cluster
> +#
> Nit: Prefer set -e on shell scripts (or explicit error checks if you prefer) to avoid silent fails.
> 
> 

OK.

> 
> In source/common/upstream/original_dst_cluster.cc <https://github.com/lyft/envoy/pull/1246#discussion_r128770391>:
> 
> > +          erased = true;
> +          break; // erase invalidates 'it'
> +        }
> +      }
> +      if (!erased) {
> +        ENVOY_LOG(debug, "REMOVING HOST {}: Not Found!", host->address()->asString());
> +      }
> +    }
> +    for (auto host : hosts_added) {
> +      ENVOY_LOG(debug, "ADDING HOST {}", host->address()->asString());
> +      auto range = host_map_.equal_range(host->address()->asString());
> +      bool found = false;
> +      for (auto it = range.first; it != range.second; ++it) {
> +        if (it->second == host) {
> +          ENVOY_LOG(debug, "ADDING HOST {}: EXISTED ALREADY!", host->address()->asString());
> +          found = true;
> Can we use std::find_if here?
> 
> 

Seems so, will do.

> 
> In source/common/upstream/original_dst_cluster.cc <https://github.com/lyft/envoy/pull/1246#discussion_r128770853>:
> 
> > +        if (localIP) {
> +          Network::Address::InstanceConstSharedPtr hostIpPort;
> +          if (localIP->version() == Network::Address::IpVersion::v4) {
> +            hostIpPort.reset(
> +                new Network::Address::Ipv4Instance(localIP->addressAsString(), localIP->port()));
> +          } else {
> +            hostIpPort.reset(
> +                new Network::Address::Ipv6Instance(localIP->addressAsString(), localIP->port()));
> +          }
> +          HostSharedPtr host = parent_.createHost(hostIpPort);
> +
> +          ENVOY_LOG(debug, "CREATED HOST {}", host->address()->asString());
> +          // Add the new host to the map.  We just failed to find it in
> +          // our local map above, so just insert it.
> +          host_map_.insert({localAddr.asString(), host});
> +          host->weight(2); // Mark as used.
> Yeah, also wondering about the magic 2 here.
> 
> In source/common/upstream/original_dst_cluster.cc <https://github.com/lyft/envoy/pull/1246#discussion_r128771920>:
> 
> > +      auto& localAddr = connection->localAddress();
> +
> +      // Check if a host with the destination address is already in the host set.
> +      auto it = host_map_.find(localAddr.asString());
> +
> +      if (it != host_map_.end()) {
> +        // Use the existing host
> +        auto host = it->second;
> +        ENVOY_LOG(debug, "FOUND HOST {}", host->address()->asString());
> +        host->weight(2); // Mark as used.
> +        return host;
> +      } else {
> +        // Add a new host
> +        auto localIP = localAddr.ip();
> +        if (localIP) {
> +          Network::Address::InstanceConstSharedPtr hostIpPort;
> How come we use the local address below as the host? I thought the idea was to use the real address from original_dst? I'm probably missing something major, so apologies if this is a silly question. A comment in the code could also be beneficial to the reader on the rationale here.
> 
> 

If connection->usingOriginalDst() is true, then the local address (destination address of the downstream connection) is the original destination address of the incoming connection. I’ll add a comment,

> 
> In source/common/json/config_schemas.cc <https://github.com/lyft/envoy/pull/1246#discussion_r128772340>:
> 
> > @@ -1305,7 +1305,7 @@ const std::string Json::Schema::CLUSTER_SCHEMA(R"EOF(
>        },
>        "type" : {
>          "type" : "string",
> -        "enum" : ["static", "strict_dns", "logical_dns", "sds"]
> +        "enum" : ["static", "strict_dns", "logical_dns", "sds", "original_dst"]
> Can we add docs on these new options in https://github.com/lyft/envoy/blob/master/docs/configuration/cluster_manager/cluster.rst <https://github.com/lyft/envoy/blob/master/docs/configuration/cluster_manager/cluster.rst> etc.
> 
> 

Will do.

  Jarno

mattklein123 commented:
@jrajahalme if possible can you respond to comments inside the GH UI? It's a lot easier to read and manage.

> Would it be bad if I changed the ClusterPtr type (which is a std::unique_ptr) to a ClusterSharedPtr? Then posts could take these and essentially hold a reference so that destruction is postponed until last of these dies off. Overall might be safer, and likely faster than passing around strings and doing hash lookup.

I think this is probably OK but I would need to think about it more. Feel free to flesh it out.

rshriram commented on source/common/upstream/original_dst_cluster.cc:
```diff
@@ -0,0 +1,163 @@
+        return host;
+      } else {
+        // Add a new host
+        auto localIP = localAddr.ip();
```
once again, would you mind eliminating the auto? I don't want this code base to degenerate into Go code. Auto is fine in super complex template types. But these are straightforward ones, I hope.

jrajahalme replied:
I was not aware of a policy on avoiding auto. Do we have one? I'm just about to post the non-RFC PR on this, but will go back and eliminate the simplest 'auto's...

rshriram replied:
I don't think there is any policy. Its more of a personal request :). I will buy you 🍺 if you get rid of the autos!

jrajahalme commented on configs/original_dst_proxy.json:
```diff
@@ -0,0 +1,44 @@
+                        "routes": [{
+                            "timeout_ms": 0,
+                            "prefix": "/",
+                            "cluster": "original_dst_cluster"
```
cluster name is not semantically important, I'll change the config example to reflect this.

jrajahalme commented:
> On Jul 24, 2017, at 6:12 AM, Shriram Rajagopalan <notifications@github.com> wrote:
> 
> @rshriram commented on this pull request.
> 
> Not sure if this PR has addressed this: we have two concepts (for the associated issue), one building on top of another.
> 
> The ability to use the original dst as the upstream destination of a connection. Requires the routes (in http, tcp) to support a orig_dst_cluster as you have done above.
> The ability to forward any tcp traffic to original dst ip and port. IOW, if there is an associated listener, use the listener. Otherwise, just forward it to requested destination as tcp connection.
> I believe the latter requires additional logic on the listener doing the SO_ORIGINAL_DST to also support a plain tcp proxy block with the original_dst_cluster as the target cluster.
> 
> 

I have not tested this with TCP proxy yet, but it seems to me that there is nothing special needed, as long as the new cluster type is used. I’ll post a new non-PR soon, maybe you can test it with TCP proxy?

  Jarno

jrajahalme commented:
Created new non-RFC PR #1314 addressing all the comments so far and building on top of the current master.

mattklein123 commented:
@jrajahalme do we need a new PR? (Seems like we could just continue iterating on this one by merging master into it). If so, can we close this one after addressing @rshriram comments in the other one?

jrajahalme commented:
> On Jul 24, 2017, at 7:47 AM, Matt Klein <notifications@github.com> wrote:
> 
> @jrajahalme <https://github.com/jrajahalme> do we need a new PR? (Seems like we could just continue iterating on this one by merging master into it). If so, can we close this one after addressing @rshriram <https://github.com/rshriram> comments in the other one?
> 
> 

OK, I’ll remove the new PR and push the new changes and master merge here. I guess I’m still a bit confused about the workflow…

  Jarno

jrajahalme commented:
The main open issue after the last round of reviews was syncing between the worker and main threads when a dynamically added cluster may be removed while membership update is being made from the worker thread, or while the posted update is queued in the master threads event queue.

Now using a weak pointer from the worker to refer to the primary cluster, which is then promoted to a shared pointer when needed. This fails if the master is gone, which is now explicitly handled. Same for the posted closure.

jrajahalme commented:
@rshriram: Added LoadBalancing context support to the TCP proxy code, *now* I don't see why it should not work :-)

mattklein123 commented:
@jrajahalme can you take a look into why the coverage test is failing? I can review once the tests are green.

jrajahalme commented:
@mattklein123 Fixed coverage. Had no idea that file scope means nothing with virtual dispatch. Wasted good 7 hours tracking this down. Live & learn, I guess...

htuch commented:
@jrajahalme are continuing the review here or will you be splitting out some more PRs?

jrajahalme commented:
> On Jul 25, 2017, at 8:26 AM, htuch <notifications@github.com> wrote:
> 
> @jrajahalme <https://github.com/jrajahalme> are we continuing the review here or will you be splitting out some more PRs?
> 
I split out the PRs you requested so far, so I would be good for review now. While reviewing, if you think additional PRs should be split out please let me know.

  Jarno

mattklein123 commented:
@htuch I can take a pass over this today.

rshriram commented on docs/configuration/cluster_manager/cluster.rst:
```diff
@@ -146,6 +147,13 @@ http2_settings
+.. _config_cluster_manager_cluster_cleanup_interval_ms:
+
+cleanup_interval_ms
+  *(optional, integer)* The interval for removing stale hosts from the cluster. If this setting is
```
would be useful to define what these "stale" hosts are.

jrajahalme replied:
Added a new commit defining "stale hosts".

mattklein123 commented on docs/configuration/cluster_manager/cluster.rst:
```diff
@@ -146,6 +147,18 @@ http2_settings
 
+.. _config_cluster_manager_cluster_cleanup_interval_ms:
+
+cleanup_interval_ms
```
Can you add this option to the JSON description above.

jrajahalme replied:
Done.

mattklein123 commented on docs/intro/arch_overview/load_balancing.rst:
```diff
@@ -54,6 +54,15 @@ The random load balancer selects a random healthy host. The random load balancer
+Original destination
+^^^^^^^^^^^^^^^^^^^^
+
+This is a special purpose load balancer that can only be used with an original destination
```
cross linking to config please

jrajahalme replied:
Added cross-linking.

mattklein123 commented on docs/intro/arch_overview/service_discovery.rst:
```diff
@@ -45,6 +45,17 @@ When interacting with large scale web services, this is the best of all possible
+Original destination
+^^^^^^^^^^^^^^^^^^^^
+
+Original destination cluster can be used when incoming connections are redirected to Envoy either
```
cross linking to config please

jrajahalme replied:
Added cross-linking.

mattklein123 commented on source/common/network/connection_impl.h:
```diff
@@ -146,6 +147,7 @@ class ConnectionImpl : public virtual Connection,
   // readDisabled(true) this allows the connection to only resume reads when readDisabled(false)
   // has been called N times.
   uint32_t read_disable_count_{0};
+  bool using_original_dst_;
```
const

jrajahalme replied:
Done.

mattklein123 commented on source/common/network/proxy_protocol.cc:
```diff
@@ -110,7 +110,7 @@ void ProxyProtocol::ActiveConnection::onReadWorker() {
   removeFromList(parent_.connections_);
 
-  listener.newConnection(fd, remote_address, local_address);
+  listener.newConnection(fd, remote_address, local_address, true);
```
shouldn't this be false?

jrajahalme replied:
Semantically this is the same case as with SO_ORIGINAL_DST, as the local address comes from the proxy protocol header, and is the original destination of the connection, rather than from the listener's listening address. Therefore 'true' is right here.

This only has effect if the request is routed to an original destination cluster, so this is fully backwards compatible, and enables use of the original destination cluster also with traffic coming via an entity adding a proxy protocol header.

mattklein123 replied:
OK makes sense.

mattklein123 commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -536,6 +538,11 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::ClusterEntry(
                                        parent.parent_.random_));
     break;
   }
+  case LoadBalancerType::OriginalDst: {
```
Can you make sure this code has coverage? (Depending on how you wrote tests it might not).

jrajahalme replied:
Checked, did already have coverage.

mattklein123 commented on source/common/upstream/cluster_manager_impl.h:
```diff
@@ -209,11 +210,11 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
   struct PrimaryClusterData {
     PrimaryClusterData(uint64_t config_hash, bool added_via_api, ClusterPtr&& cluster)
-        : config_hash_(config_hash), added_via_api_(added_via_api), cluster_(std::move(cluster)) {}
+        : config_hash_(config_hash), added_via_api_(added_via_api), cluster_(cluster.release()) {}
```
Prefer not doing release() here. Please just allocate shared_ptr at creation and pass it through. (If it's possible to get rid of ClusterPtr type entirely that would be good).

jrajahalme replied:
Done. Figured out the mock issue with shared pointers in expectations I stumbled on earlier.

Alternatively could also use a bare pointer in the factory APIs, which results in somewhat cleaner (test) code.

jrajahalme replied:
Spoke too soon. After merging with master hit another gmock issue, investigating...

jrajahalme replied:
Fixed, new patches pushed. I had converted an std::unique_ptr to a raw pointer in a factory function which then got leaked when exceptions happened. Reverted that change, all good now :-)

mattklein123 commented on source/common/upstream/original_dst_cluster.cc:
```diff
@@ -0,0 +1,145 @@
+          // OriginalDstCluster::LoadBalancer is never configured with any other type of cluster,
+          // and throws an exception otherwise.
+          Event::Dispatcher& main_dispatcher =
+              dynamic_cast<OriginalDstCluster&>(*parent).dispatcher_;
```
I'm not crazy about the use of dynamic_cast here, but I don't see a vastly better way of doing this. Can we do the dynamic_cast in the constructor though? Basically, you can just have a weak_ptr<> to your explicit type and then do the dynamic_cast once. Then you don't even need the ClusterWeakPtr definition at top level.

jrajahalme replied:
Cleaned up as proposed.

mattklein123 commented on source/common/upstream/original_dst_cluster.h:
```diff
@@ -0,0 +1,115 @@
+     */
+    class HostMap {
+    public:
+      bool insert(const HostSharedPtr& host, bool check = true) {
```
You probably already did this, but please make sure all the various interleavings here have coverage.

jrajahalme replied:
There was one "return true" block without coverage, ended up making the remove() function void and adding an ASSERT.

mattklein123 commented on test/common/upstream/original_dst_cluster_test.cc:
```diff
@@ -0,0 +1,427 @@
+#include "gtest/gtest.h"
+
+namespace Envoy {
+using testing::Invoke;
```
nit: move using outside of Envoy namespace.

jrajahalme replied:
Done.

jrajahalme commented:
> On Jul 26, 2017, at 10:38 AM, Matt Klein <notifications@github.com> wrote:
> 
> @mattklein123 commented on this pull request.
> 
> Overall very solid work. Thanks. A few random comments. @htuch <https://github.com/htuch> feel free to take a look also if you have time.
> 
> In docs/configuration/cluster_manager/cluster.rst <https://github.com/lyft/envoy/pull/1246#discussion_r129637578>:
> 
> > @@ -146,6 +147,18 @@ http2_settings
>    initiating HTTP connection pool connections. These are the same options supported in the HTTP connection
>    manager :ref:`http2_settings <config_http_conn_man_http2_settings>` option.
>  
> +.. _config_cluster_manager_cluster_cleanup_interval_ms:
> +
> +cleanup_interval_ms
> Can you add this option to the JSON description above.
> 
> 

Will do.

> 
> In docs/intro/arch_overview/load_balancing.rst <https://github.com/lyft/envoy/pull/1246#discussion_r129637856>:
> 
> > @@ -54,6 +54,15 @@ The random load balancer selects a random healthy host. The random load balancer
>  better than round robin if no health checking policy is configured. Random selection avoids bias
>  towards the host in the set that comes after a failed host.
>  
> +Original destination
> +^^^^^^^^^^^^^^^^^^^^
> +
> +This is a special purpose load balancer that can only be used with an original destination
> cross linking to config please
> 
> 

OK.

> 
> In docs/intro/arch_overview/service_discovery.rst <https://github.com/lyft/envoy/pull/1246#discussion_r129637885>:
> 
> > @@ -45,6 +45,17 @@ When interacting with large scale web services, this is the best of all possible
>  asynchronous/eventually consistent DNS resolution, long lived connections, and zero blocking in the
>  forwarding path.
>  
> +Original destination
> +^^^^^^^^^^^^^^^^^^^^
> +
> +Original destination cluster can be used when incoming connections are redirected to Envoy either
> cross linking to config please
> 
> 

Ditto.

> 
> In source/common/network/connection_impl.h <https://github.com/lyft/envoy/pull/1246#discussion_r129638317>:
> 
> > @@ -146,6 +147,7 @@ class ConnectionImpl : public virtual Connection,
>    // readDisabled(true) this allows the connection to only resume reads when readDisabled(false)
>    // has been called N times.
>    uint32_t read_disable_count_{0};
> +  bool using_original_dst_;
> const
> 
> 

OK.

> 
> In source/common/network/proxy_protocol.cc <https://github.com/lyft/envoy/pull/1246#discussion_r129638512>:
> 
> > @@ -110,7 +110,7 @@ void ProxyProtocol::ActiveConnection::onReadWorker() {
>  
>    removeFromList(parent_.connections_);
>  
> -  listener.newConnection(fd, remote_address, local_address);
> +  listener.newConnection(fd, remote_address, local_address, true);
> shouldn't this be false?
> 
> 

I guess this could be argued either way. Strictly speaking the local address is not coming from SO_ORIGNAL_DST. However, it is should not be the listener’s listening address either, as it comes from the proxy protocol header, i.e., it is the address seen on the original connection before it was diverted to envoy using the proxy protocol. I do not have a testing environment for the proxy protocol, so I’ll change this to ‘false’ unless someone speaks up.

> 
> In source/common/upstream/cluster_manager_impl.cc <https://github.com/lyft/envoy/pull/1246#discussion_r129638996>:
> 
> > @@ -536,6 +538,11 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::ClusterEntry(
>                                         parent.parent_.random_));
>      break;
>    }
> +  case LoadBalancerType::OriginalDst: {
> Can you make sure this code has coverage? (Depending on how you wrote tests it might not).
> 
> 

Will check and add test as need be.

> 
> In source/common/upstream/cluster_manager_impl.h <https://github.com/lyft/envoy/pull/1246#discussion_r129639297>:
> 
> > @@ -209,11 +210,11 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
>  
>    struct PrimaryClusterData {
>      PrimaryClusterData(uint64_t config_hash, bool added_via_api, ClusterPtr&& cluster)
> -        : config_hash_(config_hash), added_via_api_(added_via_api), cluster_(std::move(cluster)) {}
> +        : config_hash_(config_hash), added_via_api_(added_via_api), cluster_(cluster.release()) {}
> Prefer not doing release() here. Please just allocate shared_ptr at creation and pass it through. (If it's possible to get rid of ClusterPtr type entirely that would be good).
> 
> 

I started with that design, but faced too much difficulty with the cluster manager test case. Basically I could not get it right in a few hours. The issue is with the way the tests create clusters and pass them via the mock methods. I got all the tests to pass but then the test failed due to the clusters (I presume) not being freed. The test uses .release() and is thus built for the unique_ptr, and would take more refactoring than I can do at this time.

I’d suggest keeping this code as is for this PR. I can have a stab at this later, or maybe someone more familiar with the test case can have a look. It’s trivial to find the problem, as all it takes is to remove the ClusterPtr like you suggest above.

> 
> In source/common/upstream/original_dst_cluster.cc <https://github.com/lyft/envoy/pull/1246#discussion_r129642579>:
> 
> > +        host.reset(new HostImpl(info_, info_->name() + dst_addr.asString(), std::move(host_ip_port),
> +                                false, 1, ""));
> +
> +        ENVOY_LOG(debug, "Created host {}.", host->address()->asString());
> +        // Add the new host to the map.  We just failed to find it in
> +        // our local map above, so insert without checking (2nd arg == false).
> +        host_map_.insert(host, false);
> +
> +        if (ClusterSharedPtr parent = parent_.lock()) {
> +          // Dynamic cast below is guaranteed to succeed, as code instantiating the cluster
> +          // configuration, that is run prior to this code, checks that an OriginalDstCluster is
> +          // always configured with an OriginalDstCluster::LoadBalancer, and that an
> +          // OriginalDstCluster::LoadBalancer is never configured with any other type of cluster,
> +          // and throws an exception otherwise.
> +          Event::Dispatcher& main_dispatcher =
> +              dynamic_cast<OriginalDstCluster&>(*parent).dispatcher_;
> I'm not crazy about the use of dynamic_cast here, but I don't see a vastly better way of doing this. Can we do the dynamic_cast in the constructor though? Basically, you can just have a weak_ptr<> to your explicit type and then do the dynamic_cast once. Then you don't even need the ClusterWeakPtr definition at top level.
> 
> 

OK, will change.

> 
> In source/common/upstream/original_dst_cluster.h <https://github.com/lyft/envoy/pull/1246#discussion_r129642832>:
> 
> > +   */
> +  class LoadBalancer : public Upstream::LoadBalancer {
> +  public:
> +    LoadBalancer(HostSet& host_set, ClusterSharedPtr& parent);
> +
> +    // Upstream::LoadBalancer
> +    HostConstSharedPtr chooseHost(const LoadBalancerContext* context) override;
> +
> +  private:
> +    /**
> +     * Map from an host IP address/port to a HostSharedPtr.  Due to races multiple distinct host
> +     * objects with the same address can be created, so we need to use a multimap.
> +     */
> +    class HostMap {
> +    public:
> +      bool insert(const HostSharedPtr& host, bool check = true) {
> You probably already did this, but please make sure all the various interleavings here have coverage.
> 
> 

Will check coverage once more after this round of changes.

> 
> In test/common/upstream/original_dst_cluster_test.cc <https://github.com/lyft/envoy/pull/1246#discussion_r129643248>:
> 
> > +#include "common/network/address_impl.h"
> +#include "common/network/utility.h"
> +#include "common/upstream/original_dst_cluster.h"
> +#include "common/upstream/upstream_impl.h"
> +
> +#include "test/mocks/common.h"
> +#include "test/mocks/network/mocks.h"
> +#include "test/mocks/runtime/mocks.h"
> +#include "test/mocks/ssl/mocks.h"
> +#include "test/test_common/utility.h"
> +
> +#include "gmock/gmock.h"
> +#include "gtest/gtest.h"
> +
> +namespace Envoy {
> +using testing::Invoke;
> nit: move using outside of Envoy namespace.
> 
> 

OK.

  Jarno

htuch commented:
@jrajahalme It would be really helpful to reviewers if you could reply to comments using the GitHub web UI. That way, your replies get placed next to where issues are raised, rather than appearing out-of-line in large reply blobs. Thanks.

jrajahalme commented:
See replies to the comments above.

htuch commented:
@jrajahalme Thanks, much easier to follow! Unfortunately, I'm not aware of any threaded way to reply to non-review comments, GH doesn't seem great there.

jrajahalme commented on source/common/upstream/original_dst_cluster.cc:
```diff
@@ -0,0 +1,145 @@
+        } else {
+          host_ip_port.reset(
+              new Network::Address::Ipv6Instance(dst_ip->addressAsString(), dst_ip->port()));
+        }
```
I'll do a separate PR for this after this merges.

htuch replied:
That's fine. Can you add a TODO for now?

Also, can you update the title and commit message via the UI in this PR so that when we merge this we can just copy them in? Thanks.

jrajahalme replied:
Folded this in as I needed to work on this anyway, so no new TODO needed.

jrajahalme replied:
Updated the title and comment, pushing new commits to this PR in a moment.

mattklein123 commented:
@jrajahalme LGTM. Can you fix @htuch remaining comments (TODO/update title and description) and then we can get this merged?

htuch commented:
@jrajahalme Friendly ping, I think we're ready to ship if you can just adjust the tite/description/comments and merge.

mattklein123 commented:
@htuch I think this now needs a somewhat complicated master merge? That will need to get first.

jrajahalme commented:
Sorry for the silence for the week, was practically off-grid on a vacation. I'll address @htuch's comments and merge with master, should be done later today.

jrajahalme commented on include/envoy/upstream/cluster_manager.h:
```diff
@@ -164,10 +165,10 @@ class ClusterManagerFactory {
-                                     const Optional<SdsConfig>& sds_config,
-                                     Outlier::EventLoggerSharedPtr outlier_event_logger,
-                                     bool added_via_api) PURE;
+  virtual ClusterSharedPtr clusterFromJson(const Json::Object& cluster, ClusterManager& cm,
```
I've issued an envoy-api PR and will update this PR once that is merged, so that I can get the new envoy-api commit id to use in envoy/bazel/repositories.bzl.

mattklein123 commented on source/common/network/utility.cc:
```diff
@@ -134,6 +134,16 @@ Utility::parseInternetAddressAndPort(const std::string& ip_address) {
+Address::InstanceConstSharedPtr Utility::copyInternetAddressAndPort(const Address::Ip* ip) {
+  Address::InstanceConstSharedPtr addr;
+  if (ip->version() == Address::IpVersion::v4) {
+    addr.reset(new Address::Ipv4Instance(ip->addressAsString(), ip->port()));
```
nit: I would do a direct return of std::make_shared<> here and below.

jrajahalme replied:
Done.

mattklein123 commented on include/envoy/upstream/upstream.h:
```diff
@@ -95,6 +95,11 @@ class Host : virtual public HostDescription {
   virtual void weight(uint32_t new_weight) PURE;
+
+  /**
+   * @return the old value of host being in use and set the value to 'new_used'.
```
@jrajahalme I think this comment still needs attention.

jrajahalme replied:
Fixed comment, splitting to a setter/getter pair will have too wait for tomorrow.

jrajahalme replied:
Split the setter/getter in the new commits.

mattklein123 commented:
@jrajahalme LGTM. Just 2 small things and a final pass by @htuch.

jrajahalme commented on configs/original_dst_proxy_setup.sh:
```diff
@@ -0,0 +1,34 @@
+#
+# Example setup network namespace for testing Envoy original_dst cluster
+# Clean up with the cleanup script with the same arguments.
```
Can be done, but not today.

jrajahalme replied:
Done.

htuch commented on source/common/network/utility.cc:
```diff
@@ -134,6 +134,16 @@ Utility::parseInternetAddressAndPort(const std::string& ip_address) {
   return std::make_shared<Address::Ipv4Instance>(&sa4);
 }
 
+Address::InstanceConstSharedPtr Utility::copyInternetAddressAndPort(const Address::Ip* ip) {
```
Nit: Envoy style prefers `const Address::Ip& ip` here.

jrajahalme replied:
Changed.

htuch commented on source/common/network/utility.h:
```diff
@@ -67,7 +67,15 @@ class Utility {
 
   /**
    * Parse an internet host address (IPv4 or IPv6) AND port, and create an Instance from it. Throws
```
I wonder outside this PR whether we should just made this part of the `Address` interface.

mattklein123 commented:
@jrajahalme do you plan on doing any other small cleanups on this PR or are you done? (I'm fine either way).

htuch commented on configs/original-dst-cluster/README.md:
````diff
@@ -0,0 +1,76 @@
+observed from the logs:
+
+```
+bazel build --verbose_failures //source/exe:envoy-static -c dbg
````
Nit: the `--verbose_failures` is a Bazel thing, you don't need it actually for logs or anything.

jrajahalme replied:
Fixed.

htuch commented on configs/original-dst-cluster/README.md:
```diff
@@ -0,0 +1,76 @@
+# Original destination cluster configuration and testing
```
Thanks for adding this writeup! Very clear.

htuch commented:
@jrajahalme looks like some merge conflicts to resolve still.

mattklein123 commented on source/common/network/utility.h:
```diff
@@ -72,7 +72,7 @@ class Utility {
    * @param Address::Ip* to be copied to the new instance.
```
nit: `Address::Ip&`
